cli/oauth/jwt.go

```diff
@@ -0,0 +1,115 @@
+
+	// Deprecated: Email will be removed in favor of DomainClaims.Email in the
+	// future.
+	Email string `json:"email,omitempty"`
+
+	// Scope is the scopes for the claims as a string that is space delimited.
+	Scope string `json:"scope,omitempty"`
+
+	// Deprecated: SessionID should be removed as we shouldn't have sessions.
+	// For now, we cannot remove this or tutum-app will break. In the future,
+	// we'll assign session IDs and put them on the claims `jti` field.
+	SessionID string `json:"session_id,omitempty"`
+
+	// Deprecated: Source will be removed in favor of DomainClaims.Source in the
+	// future.
+	Source *Source `json:"source,omitempty"`
+
+	// Deprecated: Username will be removed in favor of DomainClaims.Username in
+	// the future.
+	Username string `json:"username,omitempty"`
+
+	// Deprecated: UserID will be removed as soon as we can.
+	UserID string `json:"user_id,omitempty"`
+}
+
+// DomainClaims represents a custom claim data set that doesn't change the spec
+// payload. This is primarily introduced by Auth0 and is defined by a fully
+// specified URL as it's key. e.g. "https://hub.docker.com"
+type DomainClaims struct {
+	// UUID is the user, machine client, or organization's UUID in our database.
+	UUID string `json:"uuid"`
+
+	// Roles is a list of roles as strings that the user has. e.g. "admin"
+	//
+	// Deprecated: This is no longer used and will be removed in the new
+	// version.
+	Roles []string `json:"roles"`
+
+	// Email is the user's email address.
+	Email string `json:"email"`
+
+	// Username is the user's username.
+	Username string `json:"username"`
+
+	// Source is the source of the JWT. This should look like
+	// `docker_{type}|{id}`.
+	Source string `json:"source"`
+
+	// SessionID is the unique ID of the token.
+	SessionID string `json:"session_id"`
+
+	// ClientID is the client_id that generated the token. This is filled if
+	// M2M.
+	ClientID string `json:"client_id,omitempty"`
+
+	// ClientName is the name of the client that generated the token. This is
+	// filled if M2M.
+	ClientName string `json:"client_name,omitempty"`
+}
+
+// Source represents a source of a JWT.
+type Source struct {
+	// Type is the type of source. This could be "pat" etc.
+	Type string `json:"type"`
+
+	// ID is the identifier to the source type. If "pat" then this will be the
+	// ID of the PAT.
+	ID string `json:"id"`
+}
+
+// ParseSigned parses a JWT and returns the signature object or error. This does
+// not verify the validity of the JWT.
+func ParseSigned(token string) (*jwt.JSONWebToken, error) {
+	return jwt.ParseSigned(token)
+}
+
+// GetClaims returns claims from an access token without verification.
+func GetClaims(accessToken string) (claims Claims, err error) {
+	token, err := ParseSigned(accessToken)
+	if err != nil {
+		return
+	}
+
+	err = token.UnsafeClaimsWithoutVerification(&claims)
+
+	return
+}
+
+func ConcatTokens(accessToken, refreshToken string) string {
+	return accessToken + ".." + refreshToken
```

Comment:
Are we sure storing the concatenated tokens under the registry URL key is the best idea?

IIRC, we talked about storing the refresh token as a separate entry in the credential store - this way older clients could still pick up the actual auth token.

Reply:
I think concatenating the tokens is valid and more useful. The oauth token will be useless anyways (1 hour validity) for older clients. I don't think we should try keep backwards compatibility like, rather migrate and force to update.

Reply:
👋 just a quick blurb; if this is what will be stored in "~/.docker/config.json", we must be very careful on the format. Updating unfortunately is not something we can require, because the file is no longer exclusive to us; there's many tools that know about the config file and that use it (kubernetes, jenkins, IDEs) .. basically; "the world", and unfortunately not all of those can be expected to be updated.

(LOL, it took 8 Years before we got rid of the old config file; and even after that, I got SPAMMED by users that we broke them 🫠🫠🫠🫠 https://github.com/docker/cli/pull/4281)

Reply:
I'm okay with not doing this (and instead storing the "usable" token only, and storing the refresh token elsewhere). 

Indeed, this rationale:
>  there's many tools that know about the config file and that use it (kubernetes, jenkins, IDEs) .. basically; "the world", and unfortunately not all of those can be expected to be updated.

Was exactly what I initially outlined as the reason to instead store the usable token there, and make sure that we always refresh that with a fresh access token on CLI startup, so that anyone accessing it (without updates) can continue using it.

However, we wouldn't be changing the "format" of the `config.json`, just what we put inside the fields we already have there (and other applications/"the world" are unlikely to be relying on credentials being there because for a lot of users these would have been stored in credential helpers/keychain/etc.), so I think we should be fine in this regard.

Reply:
Even if the access token is stored separate from the refresh token, it would be partially useless to third parties relying on it. The PAT was long lived, while the access token expires only after an hour. 

We would then need to give out a deprecation notice, even if we only store the access token in-place of the PAT in the config file.

Reply:
Yes, in a really ideal world, the plain-text store would be treated as a credentials-store, and store credentials separate from the config-file (which would also mean that for most situations the cli-config file can be immutable).

But we must be sure that whatever value we'll store in the config-file is still compatible, as it will be used by other tools.

Reply:
(it _may_ (but need to have a good look) be better / more compatible to have _NO_ credentials stored in the config than a value that breaks compatibility)

Reply:
In that kubernetes link/scenario, non-file storage is already not supported, and they suggest [creating a secret to store the credentials](https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/#create-a-secret-by-providing-credentials-on-the-command-line).

I guess we could take a couple of approaches – we can either only support non-file store scenarios with the oauth login, or we can do away with this implementation entirely and implement this on the side of the credential helpers.

I guess we could also store this under a different key (`/access-token` and `/refresh-token`), but regardless of how we do this, it won't be compatible with other clients as the value we store there will only be valid for a short amount of time.

Reply:
> (it may (but need to have a good look) be better / more compatible to have NO credentials stored in the config than a value that breaks compatibility)

I guess if we go back to using the specific keys (`index.docker.io.../access-token` and `index.docker.io.../refresh-token` this might be somewhat mitigated?